mealpy.py

```diff
@@ -132,47 +133,59 @@ def cancel_current_meal(self):
 def get_mealpal_credentials():
     config = load_config()
     email = config['email_address']
-    if config['use_keyring']:
-        password = (
-            keyring.get_password(KEYRING_SERVICENAME, email)
-            or getpass.getpass('Credential not yet stored in keychain, please enter password: ')
-        )
-        keyring.set_password(KEYRING_SERVICENAME, email, password)
-    else:
-        password = getpass.getpass('Enter password: ')
+    password = getpass.getpass('Enter password: ')
     return email, password
 
 
+def initialize_mealpal():
+    root_dir = path.abspath(path.dirname(__file__))
+    cookies_path = path.join(root_dir, COOKIES_FILENAME)
+    mealpal = MealPal()
+
+    if path.isfile(cookies_path):
+        with open(cookies_path, 'rb') as cookies_file:
+            mealpal.cookies = pickle.load(cookies_file)
+
+        # hacky way of validating cookies
+        sleep_duration = 1
+        for _ in range(5):
+            try:
+                mealpal.get_schedules('San Francisco')
+            except Exception:  # pylint: disable=broad-except
+                # Possible fluke, retry validation
+                print(f'Login using cookies failed, retrying after {sleep_duration} second(s).')
+                time.sleep(sleep_duration)
+                sleep_duration *= 2
+            else:
+                print('Login using cookies successful!')
+                return mealpal
+
+        print('Existing cookies are invalid, please re-enter your login credentials.')
+
+    while True:
+        email, password = get_mealpal_credentials()
+        if mealpal.login(email, password) == 200:
+            break
+        print('Invalid login credentials, please try again!')
+
+    # save latest cookies
+    print(f'Login successful! Saving cookies as {COOKIES_FILENAME}.')
+    with open(cookies_path, 'wb') as cookies_file:
+        pickle.dump(mealpal.cookies, cookies_file)
```

Comment:
Use `FileCookieJar` instead of pickling.

Reply:
Why not just pickle directly? `RequestsCookieJar` is a `cookielib.CookieJar`, so we need to extract the cookies into a `FileCookieJar` if we want to use that functionality.

Reply:
Pickling has security vulnerability, due to `exec`. It's why no one uses `pickle` if they only need to serialize primitives (json and yaml are more commonly used). You get the unwanted feature of unknowingly importing python objects. We can't assume the user will be responsible for managing the file perms to ensure that only they have write access, that seems negligent.

There's already a standard serialization format, using `pickle` to circumvent that is akin to "rolling your own serialization". Btw, `FileCookieJar` outputs to a standard cookie.txt format, so it's import/exportable from a browser (from what i've read).

If i'm reading the `requests` docs  correctly, you only need to instantiate a `CookieJar` and pass it to `requests` session object. Otherwise it's instantiating a `RequestsCookieJar` on your behalf.

Reply:
Thanks for sharing about `pickle`! I've updated the code to use sessions along with `MozillaCookieJar` that is able to load and save cookies directly.

Reply:
Sources for further reading about all the dangers of deserialization:
- https://www.kevinlondon.com/2015/07/26/dangerous-python-functions.html
- https://security.openstack.org/guidelines/dg_avoid-dangerous-input-parsing-libraries.html

I used `strictyaml` instead of the standard `PyYaml` (`yaml`) for config reading. This is because`strictyaml` is [restricted yaml implementation](https://hitchdev.com/strictyaml/features-removed/).